kotlinpoet/src/jvmMain/kotlin/com/squareup/kotlinpoet/Util.jvm.kt

```diff
@@ -0,0 +1,111 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.squareup.kotlinpoet
+
+import com.squareup.kotlinpoet.CodeBlock.Companion.isPlaceholder
+import java.util.Collections
+import java.util.EnumSet
+import java.util.TreeSet
+import kotlin.collections.toSortedSet as toSortedSetKt
+import kotlin.reflect.KClass
+import kotlin.sequences.toSortedSet as toSortedSetKt
+
+// TODO Wait for CodeBlock
+internal fun CodeBlock.ensureEndsWithNewLine() = trimTrailingNewLine('\n')
+
+internal fun CodeBlock.trimTrailingNewLine(replaceWith: Char? = null) = if (isEmpty()) {
+  this
+} else {
+  with(toBuilder()) {
+    val lastFormatPart = trim().formatParts.last()
+    if (lastFormatPart.isPlaceholder && args.isNotEmpty()) {
+      val lastArg = args.last()
+      if (lastArg is String) {
+        val trimmedArg = lastArg.trimEnd('\n')
+        args[args.size - 1] = if (replaceWith != null) {
+          trimmedArg + replaceWith
+        } else {
+          trimmedArg
+        }
+      }
+    } else {
+      formatParts[formatParts.lastIndexOf(lastFormatPart)] = lastFormatPart.trimEnd('\n')
+      if (replaceWith != null) {
+        formatParts += "$replaceWith"
+      }
+    }
+    return@with build()
+  }
+}
+
+// TODO Wait for KModifier
+internal fun requireNoneOrOneOf(modifiers: Set<KModifier>, vararg mutuallyExclusive: KModifier) {
+  val count = mutuallyExclusive.count(modifiers::contains)
+  require(count <= 1) {
+    "modifiers $modifiers must contain none or only one of ${mutuallyExclusive.contentToString()}"
+  }
+}
+
+internal fun requireNoneOf(modifiers: Set<KModifier>, vararg forbidden: KModifier) {
+  require(forbidden.none(modifiers::contains)) {
+    "modifiers $modifiers must contain none of ${forbidden.contentToString()}"
+  }
+}
+
+internal actual fun formatIsoControlCode(code: Int): String =
+  String.format("\\u%04x", code)
+
+internal actual fun Int.toHexStr(): String =
+  Integer.toHexString(this)
+
+internal actual fun <K, V> Map<K, V>.toImmutableMap(): Map<K, V> =
+  Collections.unmodifiableMap(LinkedHashMap(this))
+
+internal actual fun <T> Collection<T>.toImmutableList(): List<T> =
+  Collections.unmodifiableList(ArrayList(this))
+
+internal actual fun <T> Collection<T>.toImmutableSet(): Set<T> =
+  Collections.unmodifiableSet(LinkedHashSet(this))
+
+internal actual fun <T : Comparable<T>> Sequence<T>.toSortedSet(): Set<T> =
+  toSortedSetKt()
+
+internal actual fun <T : Comparable<T>> List<T>.toSortedSet(): Set<T> =
+  toSortedSetKt()
+
+internal actual fun <T : Comparable<T>> sortedSetOf(): MutableSet<T> =
+  TreeSet()
+
+internal actual inline fun <reified E : Enum<E>> enumSetOf(vararg values: E): MutableSet<E> {
```

Comment:
Where is this function used? Why does the implementation need a `when`, can we implement it in the same way `toEnumSet` was implemented?

Reply:
`enumSetOf` is mainly used to replace the previous direct use of `EnumSet.of(...)`, compared to `toEnumSet`, in the JVM may be a little more efficient? But it's not very efficient, so `toEnumSet` is fine.
The use of `when` is also only because it avoids a `copyOf` copy, so there may be a small efficiency gain, but it's not much.

Reply:
Got it, fine with leaving as is.